### PR TITLE
Add Codeartifact resources 

### PR DESCRIPTION
**Pull request by stevendborrelli**
### Description of your changes

Adds [AWS CodeArtifact](https://aws.amazon.com/codeartifact/) resources. 

- <https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/codeartifact_domain>
- <https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/codeartifact_domain_permissions_policy>
- <https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/codeartifact_repository>
- <https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/codeartifact_repository_permissions_policy>

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

All resources tested in uptest <https://github.com/crossplane-contrib/provider-upjet-aws/actions/runs/9618965896>

Local run for Repository/Domain/RepositoryPermissionsPolicy:

```shell
kubectl get -f rpp.yaml 
NAME                                                                                                 SYNCED   READY   EXTERNAL-NAME                                                                                                                                      AGE
repositorypermissionspolicy.codeartifact.aws.upbound.io/repositorypermissionspolicy-example-policy   True     True    arn:aws:codeartifact:us-east-1:609897127049:repository/repositorypermissionspolicy-example-domain/repositorypermissionspolicy-example-repository   35s

NAME                                                                            SYNCED   READY   EXTERNAL-NAME                                AGE
domain.codeartifact.aws.upbound.io/repositorypermissionspolicy-example-domain   True     True    repositorypermissionspolicy-example-domain   34s

NAME                                                                                    SYNCED   READY   EXTERNAL-NAME                                    AGE
repository.codeartifact.aws.upbound.io/repositorypermissionspolicy-example-repository   True     True    repositorypermissionspolicy-example-repository   34s
```

Local run for Domain/DomainPolicy:

```shell
NAME                                                                                  SYNCED   READY   EXTERNAL-NAME                                                                        AGE
domainpermissionspolicy.codeartifact.aws.upbound.io/domainpermissionspolicy-example   True     True    arn:aws:codeartifact:us-east-1:609897127049:domain/domainpermissionspolicy-example   10s

NAME                                                                 SYNCED   READY   EXTERNAL-NAME                     AGE
domain.codeartifact.aws.upbound.io/domainpermissionspolicy-example   True     True    domainpermissionspolicy-example   10s

```

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by stevendborrelli** [comment]
/test-examples="examples/codeartifact/v1beta1/domain.yaml"

**Comment by stevendborrelli** [comment]
/test-examples="examples/codeartifact/v1beta1/domain.yaml,examples/codeartifact/v1beta1/respository.yaml"

**Comment by stevendborrelli** [comment]
/test-examples="examples/codeartifact/v1beta1/domainpermissionspolicy.yaml"

**Comment by stevendborrelli** [comment]
/test-examples="examples/codeartifact/v1beta1/repositorypermissionspolicy.yaml"

File: config/externalname.go [hunk 1]
```diff
@@ -154,6 +154,17 @@ var TerraformPluginSDKExternalNameConfigs = map[string]config.ExternalName{
 	// Event data stores can be imported using their arn
 	"aws_cloudtrail_event_data_store": config.IdentifierFromProvider,
 
+	// codeartifact
+	//
+	// CodeArtifact Domain can be imported using the CodeArtifact Domain arn
+	"aws_codeartifact_domain": config.IdentifierFromProvider,
+	// CodeArtifact Domain Permissions Policies can be imported using the CodeArtifact Domain ARN
+	"aws_codeartifact_domain_permissions_policy": config.IdentifierFromProvider,
```
Comment (mbbush) [inline config/externalname.go:162]: If we can, it's better to construct the external name from the `spec`. It looks like with at least the domain_permissions_policy and repository_permissions_policy that should be easy enough to do.

Comment (stevendborrelli) [reply config/externalname.go:162]: @mbbush do you mean something like https://github.com/crossplane-contrib/provider-upjet-aws/blob/87d323b06a409d034515026db83c87333c34cad6/config/externalname.go#L106? 

```
        "aws_prometheus_rule_group_namespace": config.TemplatedStringAsIdentifier("name", "arn:aws:aps:{{ .setup.configuration.region }}:{{ .setup.client_metadata.account_id }}:rulegroupsnamespace/{{ .parameters.workspace_id }}/{{ .external_name }}"),
```

Comment (stevendborrelli) [reply config/externalname.go]: @mbbush I've updated the code

**Comment by stevendborrelli** [comment]
/test-examples="examples/codeartifact/v1beta1/domainpermissionspolicy.yaml,examples/codeartifact/v1beta1/repositorypermissionspolicy.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/codeartifact/v1beta1/domainpermissionspolicy.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/codeartifact/v1beta1/repository.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/codeartifact/v1beta1/repositorypermissionspolicy.yaml"

File: apis/codeartifact/v1beta1/reference.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+import (
+	"github.com/crossplane/crossplane-runtime/pkg/reference"
+	xpresource "github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ExtractDomainName extracts the name from a Domain
+func ExtractDomainName() reference.ExtractValueFn {
```
Comment (turkenf) [inline apis/codeartifact/v1beta1/reference.go:15]: We have the [ExtractParamPath](https://github.com/crossplane/upjet/blob/37c7f4e91d57982f58e496dfb86dd7c2d29a1921/pkg/resource/extractor.go#L34) function in upjet to extract the desired parameters. I will leave a comment on the reference extractors below. If it works, we can remove the `reference.go` file. What do you think?

File: config/codeartifact/config.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+		r.References["encryption_key"] = config.Reference{
+			TerraformName: "aws_kms_key",
+			Extractor:     common.PathARNExtractor,
+		}
+	})
+
+	p.AddResourceConfigurator("aws_codeartifact_domain_permissions_policy", func(r *config.Resource) {
+		r.References["domain"] = config.Reference{
+			TerraformName: "aws_codeartifact_domain",
+			Extractor:     "ExtractDomainName()",
```
Comment (turkenf) [inline config/codeartifact/config.go:25]: We can use the ExtractParamPath function I mentioned in the comment above.

```suggestion
			TerraformName: "aws_codeartifact_domain",
			Extractor:     `github.com/crossplane/upjet/pkg/resource.ExtractParamPath("domain",true)`,
```

As an example: https://github.com/crossplane-contrib/provider-upjet-aws/blob/4dd9c70f4fd92615db469a1715d14f79522b5368/config/kafkaconnect/config.go#L35

File: config/codeartifact/config.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+		}
+	})
+
+	p.AddResourceConfigurator("aws_codeartifact_repository", func(r *config.Resource) {
+		r.References["domain"] = config.Reference{
+			TerraformName: "aws_codeartifact_domain",
+			Extractor:     "ExtractDomainName()",
```
Comment (turkenf) [inline config/codeartifact/config.go:32]:
```suggestion
			Extractor:    `github.com/crossplane/upjet/pkg/resource.ExtractParamPath("domain",true)`,
```

File: config/codeartifact/config.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+		}
+	})
+
+	p.AddResourceConfigurator("aws_codeartifact_repository_permissions_policy", func(r *config.Resource) {
+		r.References["domain"] = config.Reference{
+			TerraformName: "aws_codeartifact_domain",
+			Extractor:     "ExtractDomainName()",
```
Comment (turkenf) [inline config/codeartifact/config.go:39]:
```suggestion
			Extractor:     `github.com/crossplane/upjet/pkg/resource.ExtractParamPath("domain",true)`,
```

File: config/codeartifact/config.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+		}
+
+		r.References["repository"] = config.Reference{
+			TerraformName: "aws_codeartifact_repository",
+			Extractor:     "ExtractRepositoryName()",
```
Comment (turkenf) [inline config/codeartifact/config.go:44]:
```suggestion
			Extractor:     `github.com/crossplane/upjet/pkg/resource.ExtractParamPath("repository",true)`,
```

**Review by turkenf — COMMENTED** [review]
Thank you for your effort in adding these resources @stevendborrelli, I left a few comments about [`ExtractParamPath`](https://github.com/crossplane/upjet/blob/37c7f4e91d57982f58e496dfb86dd7c2d29a1921/pkg/resource/extractor.go#L34) function for you to consider. 

If this function works for us, I would prefer to continue with `ExtractParamPath`. You can ignore my comments if you have already tried it and it did not work.

File: apis/codeartifact/v1beta1/reference.go (hunk 2 above)
Comment (stevendborrelli) [reply apis/codeartifact/v1beta1/reference.go:15]: I'll make the changes and ping you when it is ready for review.

File: config/codeartifact/config.go (hunk 6 above)
Comment (stevendborrelli) [reply config/codeartifact/config.go]: Updated in https://github.com/crossplane-contrib/provider-upjet-aws/pull/1368/commits/2f037e2a74d6ccbc4a0d79fb4239b8ffc682a9e5

File: config/codeartifact/config.go (hunk 5 above)
Comment (stevendborrelli) [reply config/codeartifact/config.go]: Updated in https://github.com/crossplane-contrib/provider-upjet-aws/pull/1368/commits/2f037e2a74d6ccbc4a0d79fb4239b8ffc682a9e5

**Comment by stevendborrelli** [comment]
/test-examples="examples/codeartifact/v1beta1/domain.yaml,examples/codeartifact/v1beta1/respository.yaml"

**Comment by stevendborrelli** [comment]
/test-examples="examples/codeartifact/v1beta1/repositorypermissionspolicy.yaml"

**Comment by stevendborrelli** [comment]
/test-examples="examples/codeartifact/v1beta1/domainpermissionspolicy.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/codeartifact/v1beta1/repository.yaml"

**Review by turkenf — APPROVED** [review]
Thank you @stevendborrelli, LGTM.

**Pull request merged**